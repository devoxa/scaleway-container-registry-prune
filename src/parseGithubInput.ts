const UUID_REGEX = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-5][0-9a-f]{3}-[089ab][0-9a-f]{3}-[0-9a-f]{12}$/i
const IMAGE_REGEX = /^rg\.(.*?)\.scw.cloud\/[^/]+\/([^/]+)$/i

export interface GithubInput {
  scwSecretToken: string
  image: string
  tagPattern: string
  keepLast: string
}

export interface Options {
  scwSecretToken: string
  region: string
  imageName: string
  tagPattern: RegExp
  keepLast: number
}

export function parseGithubInput(input: GithubInput): Options {
  if (!input.scwSecretToken.match(UUID_REGEX)) {
    throw new Error('\'scw-secret-token\' is not a valid "secret key" part of a Scaleway API key')
  }

  const imageMatch = input.image.match(IMAGE_REGEX)
  if (!imageMatch) {
    throw new Error("'image' is not the name of an image on the Scaleway Container Registry")
  }
  const region = imageMatch[1]
  const imageName = imageMatch[2]

  let tagPattern
  try {
    tagPattern = new RegExp(input.tagPattern)
  } catch (err) {
    if (!(err instanceof Error)) throw err
    throw new Error("'tag-pattern' is not a valid regular expression")
  }

  if (!input.keepLast.match(/^\d+$/)) {
    throw new Error("'keep-last' is not a valid positive integer")
  }

  const keepLast = parseInt(input.keepLast, 10)
  if (keepLast >= 100) {
    throw new Error("'keep-last' must be smaller than 100")
  }

  return {
    scwSecretToken: input.scwSecretToken,
    region,
    imageName,
    tagPattern,
    keepLast,
  }
}
