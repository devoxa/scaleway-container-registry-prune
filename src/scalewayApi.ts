import querystring from 'querystring'

type ScwGetImagesResponse = { images?: Array<ScwImage> }
type ScwImage = { id: string; name: string }

type ScwGetTagsResponse = { tags?: Array<ScwTag> }
type ScwTag = { id: string; name: string; updated_at: string }

type ScwDeleteTagResponse = { id: string; name: string; updated_at: string }

export async function getImage(
  scwSecretToken: string,
  region: string,
  imageName: string
): Promise<ScwImage> {
  const response = await fetch(
    `https://api.scaleway.com/registry/v1/regions/${region}/images?name=${imageName}`,
    { headers: { 'x-auth-token': scwSecretToken } }
  )

  const json = (await response.json()) as ScwGetImagesResponse

  if (!json.images || !json.images[0]) {
    throw new Error('Could not get image from the API')
  }

  return json.images[0]
}

export async function listTags(
  scwSecretToken: string,
  region: string,
  imageId: string
): Promise<Array<ScwTag>> {
  const query = querystring.encode({
    page_size: 100,
  })

  const response = await fetch(
    `https://api.scaleway.com/registry/v1/regions/${region}/images/${imageId}/tags?${query}`,
    { headers: { 'x-auth-token': scwSecretToken } }
  )

  const json = (await response.json()) as ScwGetTagsResponse

  if (!json.tags || json.tags.length === 0) {
    throw new Error('Could not list tags from the API')
  }

  // Sort the tags by the time they were last updated (most recent first)
  const tags = json.tags
  tags.sort((a, b) => new Date(b.updated_at).getTime() - new Date(a.updated_at).getTime())

  return tags
}

export async function deleteTag(
  scwSecretToken: string,
  region: string,
  tagId: string
): Promise<ScwDeleteTagResponse> {
  const query = querystring.encode({
    force: true,
  })

  const response = await fetch(
    `https://api.scaleway.com/registry/v1/regions/${region}/tags/${tagId}?${query}`,
    { method: 'DELETE', headers: { 'x-auth-token': scwSecretToken } }
  )

  const json = (await response.json()) as ScwDeleteTagResponse

  if (json.id !== tagId) {
    throw new Error('Could not delete tag from the API')
  }

  return json
}
