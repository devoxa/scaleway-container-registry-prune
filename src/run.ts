import { GithubInput, parseGithubInput } from './parseGithubInput'
import { deleteTag, getImage, listTags } from './scalewayApi'

export async function run(input: GithubInput): Promise<void> {
  const options = parseGithubInput(input)

  const image = await getImage(options.scwSecretToken, options.region, options.imageName)
  const tags = await listTags(options.scwSecretToken, options.region, image.id)

  let count = 0
  for (const tag of tags) {
    if (!tag.name.match(options.tagPattern)) {
      console.log(`Skipping tag: ${tag.name} (${tag.updated_at})`)
      continue
    }

    count++

    if (count <= options.keepLast) {
      console.log(`Keeping tag: ${tag.name} (${tag.updated_at})`)
      continue
    }

    console.log(`Pruning tag: ${tag.name} (${tag.updated_at})`)
    await deleteTag(options.scwSecretToken, options.region, tag.id)
  }
}
