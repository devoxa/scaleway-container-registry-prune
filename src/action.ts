import { getInput } from '@actions/core'
import { run } from './run'

run({
  scwSecretToken: getInput('scw-secret-token'),
  image: getInput('image'),
  tagPattern: getInput('tag-pattern'),
  keepLast: getInput('keep-last'),
}).catch((err) => {
  console.log('Error: ' + err.message)
  process.exit(1)
})
