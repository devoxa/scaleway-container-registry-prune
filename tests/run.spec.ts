import { run } from '../src/run'
import * as scalewayApi from '../src/scalewayApi'

const logSpy = jest.spyOn(console, 'log')

jest.mock('../src/scalewayApi')
const mockGetImage = jest.mocked(scalewayApi.getImage)
const mockListTags = jest.mocked(scalewayApi.listTags)
const mockDeleteTag = jest.mocked(scalewayApi.deleteTag)

const INPUT = {
  scwSecretToken: '03a23c20-d00c-4e80-9166-bf9318c949de',
  image: 'rg.nl-ams.scw.cloud/devoxa/hello-world',
  tagPattern: '^(?!pr-).+$',
  keepLast: '2',
}

describe('run', () => {
  beforeEach(() => {
    logSpy.mockReset()
    mockGetImage.mockReset()
    mockListTags.mockReset()
    mockDeleteTag.mockReset()
  })

  test('can prune excess tags', async () => {
    mockGetImage.mockReturnValueOnce(Promise.resolve({ id: 'image-id', name: 'image-name' }))
    mockListTags.mockReturnValueOnce(
      Promise.resolve([
        { id: 'tag-id-4', name: 'tag-name-a', updated_at: '2021-01-04' }, // Keep
        { id: 'tag-id-3', name: 'pr-123', updated_at: '2021-01-03' }, // Skip
        { id: 'tag-id-2', name: 'tag-name-b', updated_at: '2021-01-02' }, // Keep
        { id: 'tag-id-1', name: 'tag-name-c', updated_at: '2021-01-01' }, // Prune
        { id: 'tag-id-0', name: 'tag-name-d', updated_at: '2020-01-01' }, // Prune
      ])
    )

    await run(INPUT)

    expect(logSpy.mock.calls).toMatchSnapshot()
    expect(mockDeleteTag.mock.calls).toMatchSnapshot()
  })

  test('can prune all tags', async () => {
    mockGetImage.mockReturnValueOnce(Promise.resolve({ id: 'image-id', name: 'image-name' }))
    mockListTags.mockReturnValueOnce(
      Promise.resolve([
        { id: 'tag-id-4', name: 'tag-name-a', updated_at: '2021-01-04' }, // Prune
        { id: 'tag-id-3', name: 'pr-123', updated_at: '2021-01-03' }, // Skip
        { id: 'tag-id-2', name: 'tag-name-b', updated_at: '2021-01-02' }, // Prune
        { id: 'tag-id-1', name: 'tag-name-c', updated_at: '2021-01-01' }, // Prune
        { id: 'tag-id-0', name: 'tag-name-d', updated_at: '2020-01-01' }, // Prune
      ])
    )

    await run({ ...INPUT, keepLast: '0' })

    expect(mockDeleteTag.mock.calls).toMatchSnapshot()
  })

  test('can prune no tags', async () => {
    mockGetImage.mockReturnValueOnce(Promise.resolve({ id: 'image-id', name: 'image-name' }))
    mockListTags.mockReturnValueOnce(
      Promise.resolve([
        { id: 'tag-id-4', name: 'tag-name-a', updated_at: '2021-01-04' }, // Keep
        { id: 'tag-id-3', name: 'pr-123', updated_at: '2021-01-03' }, // Skip
        { id: 'tag-id-2', name: 'tag-name-b', updated_at: '2021-01-02' }, // Keep
        { id: 'tag-id-1', name: 'tag-name-c', updated_at: '2021-01-01' }, // Keep
        { id: 'tag-id-0', name: 'tag-name-d', updated_at: '2020-01-01' }, // Keep
      ])
    )

    await run({ ...INPUT, keepLast: '50' })

    expect(mockDeleteTag.mock.calls).toEqual([])
  })
})
