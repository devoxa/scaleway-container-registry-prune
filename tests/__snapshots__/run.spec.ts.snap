// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`run can prune all tags 1`] = `
[
  [
    "03a23c20-d00c-4e80-9166-bf9318c949de",
    "nl-ams",
    "tag-id-4",
  ],
  [
    "03a23c20-d00c-4e80-9166-bf9318c949de",
    "nl-ams",
    "tag-id-2",
  ],
  [
    "03a23c20-d00c-4e80-9166-bf9318c949de",
    "nl-ams",
    "tag-id-1",
  ],
  [
    "03a23c20-d00c-4e80-9166-bf9318c949de",
    "nl-ams",
    "tag-id-0",
  ],
]
`;

exports[`run can prune excess tags 1`] = `
[
  [
    "Keeping tag: tag-name-a (2021-01-04)",
  ],
  [
    "Skipping tag: pr-123 (2021-01-03)",
  ],
  [
    "Keeping tag: tag-name-b (2021-01-02)",
  ],
  [
    "Pruning tag: tag-name-c (2021-01-01)",
  ],
  [
    "Pruning tag: tag-name-d (2020-01-01)",
  ],
]
`;

exports[`run can prune excess tags 2`] = `
[
  [
    "03a23c20-d00c-4e80-9166-bf9318c949de",
    "nl-ams",
    "tag-id-1",
  ],
  [
    "03a23c20-d00c-4e80-9166-bf9318c949de",
    "nl-ams",
    "tag-id-0",
  ],
]
`;
